drivers: sensor: Add driver for SB-TSI

panikiel opened the pull request:
Add a driver for the SB Temperature Sensor Interface. This is an I2C temperature sensor on AMD SoCs.

keith-zephyr commented on drivers/sensor/amd_sb_tsi/sb_tsi_emul.h:
See my comments on https://github.com/zephyrproject-rtos/zephyr/pull/60833.  Please convert this emulator to use the common backend support.

semihalf-anikiel-pawel replied:
Done. `tests/drivers/build_all/sensor/sensors.generic_test` indeed runs my emulator (after manually setting `build_only` to false)

tristan-google replied:
> after manually setting build_only to false)

Yeah, the generic test is currently disabled upstream due to an unrelated driver acting up in CI. I'm working on addressing this.

MaureenHelm commented on drivers/sensor/amd_sb_tsi/sb_tsi.c:
```diff
@@ -0,0 +1,88 @@
+static int sb_tsi_sample_fetch(const struct device *dev,
+			       enum sensor_channel chan)
+{
+	struct sb_tsi_data *drv_data = dev->data;
+	const struct sb_tsi_config *cfg = dev->config;
```
Convention is to call these local variables `data` and `config` respectively. Please fix throughout

semihalf-anikiel-pawel replied:
Done

MaureenHelm commented on drivers/sensor/amd_sb_tsi/sb_tsi.h:
```diff
@@ -0,0 +1,27 @@
+#include <zephyr/device.h>
+#include <zephyr/sys/util.h>
+
+#define SB_TSI_TEMP_INT 0x01
+#define SB_TSI_TEMP_DEC 0x10
+#define SB_TSI_TEMP_DEC_SHIFT 5
+#define SB_TSI_TEMP_DEC_SCALE 8
+
+struct sb_tsi_data {
+	uint8_t sample_int;
+	uint8_t sample_dec;
+};
+
+struct sb_tsi_config {
+	struct i2c_dt_spec i2c;
+};
```
There's not enough here to justify a separate private header file; please move these definitions to the .c file

tristan-google commented on drivers/sensor/amd_sb_tsi/sb_tsi_emul.c:
```diff
@@ -0,0 +1,157 @@
+		return -ENOTSUP;
+	}
+
+	scaled_value = (int64_t)value << shift;
+	temp_eighths = scaled_value >> 28;
+	temp_eighths = CLAMP(temp_eighths, 0, 0x7ff);
+
+	data->reg[SB_TSI_TEMP_INT] = temp_eighths >> 3;
+	data->reg[SB_TSI_TEMP_DEC] = (temp_eighths & 0x7) << 5;
```
Continuing on from the other comment, what you want to do here is convert the fixed-point Celsius value passed in by the test to its corresponding register representation. In this case I'd unshift the supplied value and apply the appropriate scaling factor to convert to integer millicelsius. Then divide by `1/0.125 = 8` and clamp to get the register value.

semihalf-anikiel-pawel replied:
Is this better?

tristan-google commented on drivers/sensor/amd_sb_tsi/sb_tsi_emul.c:
```diff
@@ -0,0 +1,157 @@
+		return -ENOTSUP;
+	}
+
+	*lower = 0;
```
It looks like you're using these fields to pass around a raw register representation of the temperature rather than degrees Celsius in fixed-point format, which is the encoding the new sensors API as well as the backend emul API uses in place of the traditional `struct sensor_val`.

Based on what I could find this sensor runs a range of 0.125 C to 255.875 C. This will fit within the range of shift value 8 (Q9.23 in DSP notation), which is `-256` to `255.9999998807907`. The sensor's lower bound (0.125 C), for example, would be encoded as `0.125 * 2^31 >> 8 = 0x100000` (but I would just write the full expression instead of pasting the hex value for readability's sake: `*lower = (int64_t)(0.125 * ((int64_t)INT32_MAX + 1)) >> *shift;`). In this case, epsilon would be the same value because we go up in increments of 0.125 C. Epsilon is used as a tolerance when the generic test compares actual to expected to allow for rounding / precision errors. The value of `*upper` would be `0x7ff00000` (same math as above)

Let me know if you have any other questions, this can be hard to wrap one's head around at first. I found https://chummersone.github.io/qformat.html very useful. (For this case you'd plug in 32 as word size and 23 as the number of fractional bits)

semihalf-anikiel-pawel replied:
Thank you very much for the explanation. I applied the changes, except for the lower bound being 0C, not 0.125C

Commit:
drivers: sensor: Add driver for SB-TSI

Add a driver for the SB Temperature Sensor Interface. This is an I2C
temperature sensor on AMD SoCs.

Signed-off-by: Paweł Anikiel <pan@semihalf.com>

keith-zephyr commented on drivers/sensor/amd_sb_tsi/sb_tsi.c:
```diff
@@ -0,0 +1,118 @@
+		return -ENOTSUP;
+	}
+
+	res = pm_device_runtime_get(dev);
+	if (res) {
+		return res;
+	}
```
I don't see other sensors using power management this way.  There are few examples of sensors using the power management action to put the sensor into a low power mode.  But what you've implemented here would prevent a power domain from going into suspend.
[bmi323 example](https://github.com/zephyrproject-rtos/zephyr/blob/762a1f8b42412c24f76b1533df69fc5d27b8b70a/drivers/sensor/bmi323/bmi323.c#L1235)

@teburd - you mentioned defining a PM strategy for sensors [here](https://github.com/zephyrproject-rtos/zephyr/pull/55790#issuecomment-1472001677).  What do you think?

semihalf-anikiel-pawel replied:
I changed it to be like how it's done in bmi160 (as linked above).

keith-zephyr commented on drivers/sensor/amd_sb_tsi/sb_tsi_emul.c:
```diff
@@ -0,0 +1,158 @@
+};
+
+#define SB_TSI_EMUL(n)								\
+	const struct sb_tsi_emul_cfg sb_tsi_emul_cfg_##n;			\
```
See above.  const `config` info is not required.

semihalf-anikiel-pawel replied:
Done

keith-zephyr commented on drivers/sensor/amd_sb_tsi/sb_tsi_emul.c:
```diff
@@ -0,0 +1,158 @@
+	const struct sb_tsi_emul_cfg sb_tsi_emul_cfg_##n;			\
+	struct sb_tsi_emul_data sb_tsi_emul_data_##n;				\
+	EMUL_DT_INST_DEFINE(n, sb_tsi_emul_init, &sb_tsi_emul_data_##n,		\
+			    &sb_tsi_emul_cfg_##n, &sb_tsi_emul_api_i2c,		\
```
```suggestion
			    NULL, &sb_tsi_emul_api_i2c,		\
```

semihalf-anikiel-pawel replied:
Done

semihalf-anikiel-pawel commented on drivers/sensor/amd_sb_tsi/sb_tsi_emul.c:
```diff
@@ -0,0 +1,158 @@
+};
+
+struct sb_tsi_emul_cfg {
+};
```
Done

yperess commented on drivers/sensor/amd_sb_tsi/sb_tsi.c:
```diff
@@ -0,0 +1,131 @@
+		return -EIO;
+	}
+
+	if (chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_DIE_TEMP) {
```
I think this is the wrong channel, die temperature is usually the internal temperature of the sensor used for calibration. I think you want `SENSOR_CHAN_AMBIENT_TEMP`

semihalf-anikiel-pawel replied:
Done

yperess commented on drivers/sensor/amd_sb_tsi/sb_tsi.c:
```diff
@@ -0,0 +1,131 @@
+{
+	struct sb_tsi_data *data = dev->data;
+
+	if (chan != SENSOR_CHAN_DIE_TEMP) {
```
Same here, use `SENSOR_CHAN_AMBIENT_TEMP`

semihalf-anikiel-pawel replied:
Done

semihalf-anikiel-pawel commented:
@yperess, could you please take another look?
